Fix Asynchronous PyTorch Profiler Trace

sraikund16 opened the pull request:
Summary: With the merge of D55925068, we have introduced an overflow issue when recording a trace using dyno gputrace. This is because it is possible for TorchOPs to be enumerated but not have an end time since they were running as the recording ended. By default these events have an end time set to INT_MIN. When finding the duration() for such events using end-start, we get an overflow resulting in a very long duration. This was avoided before because we were dividing the INT_MIN by 1000 because we were trying to convert uS to nS. This change introduces a patch for TorchOps and a future PR will be added to create a more universal guard in kineto.

Test Plan:
Trace recorded using resnet test.

Trace:
https://www.internalfb.com/intern/perfdoctor/trace_view?filepath=tree/traces/dynocli/0/1713199267/localhost/libkineto_activities_2247224.json.gz&bucket=gpu_traces

Differential Revision: D56144914

aaronenyeshi reviewed (approved):
LGTM, we should add a little more detail about why INT_MIN's conversion into a large positive number won't cause std::max to always choose that large number.

sraikund16 committed:
Fix Asynchronous PyTorch Profiler Trace (#124080)

Summary:

With the merge of D55925068, we have introduced an overflow issue when recording a trace using dyno gputrace. This is because it is possible for TorchOPs to be enumerated but not have an end time since they were running as the recording ended. By default these events have an end time set to INT_MIN. When finding the duration() for such events using end-start, we get an overflow resulting in a very long duration. This was avoided before because we were dividing the INT_MIN by 1000 because we were trying to convert uS to nS. This change introduces a patch for TorchOps and a future PR will be added to create a more universal guard in kineto.

Test Plan:
Trace recorded using resnet test.

Trace:
https://www.internalfb.com/intern/perfdoctor/trace_view?filepath=tree/traces/dynocli/0/1713199267/localhost/libkineto_activities_2247224.json.gz&bucket=gpu_traces

Reviewed By: aaronenyeshi

Differential Revision: D56144914

sraikund16 commented:
We are picking between the max of two numbers because essentially if end_time is less than start_time we can just set it to start_time. This will result in durations that will always be >=0. The overflow that is being caused is not from an improper casting, it is because we subtract INT_MIN from a positive number in kineto. In this case, we will always have one positive number subtracting from another positive number so no overflow.
